lib/chat_api_web/controllers/canned_response_controller.ex

```diff
@@ -0,0 +1,84 @@
+defmodule ChatApiWeb.CannedResponseController do
+  use ChatApiWeb, :controller
+
+  alias ChatApi.CannedResponses
+  alias ChatApi.CannedResponses.CannedResponse
+
+  action_fallback ChatApiWeb.FallbackController
+
+  @spec index(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def index(conn, _params) do
+    with %{account_id: account_id} <- conn.assigns.current_user do
+      canned_responses = CannedResponses.list_canned_responses_by_account(account_id)
+      render(conn, "index.json", canned_responses: canned_responses)
+    end
+  end
+
+  @spec create(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def create(
+        conn,
+        %{
+          "content" => content,
+          "name" => name
+        }
+      ) do
+    with %{account_id: account_id} <- conn.assigns.current_user,
+         {:ok, %CannedResponse{} = canned_response} <-
+           CannedResponses.create_canned_response(%{
+             content: content,
+             name: name,
+             account_id: account_id
+           }) do
+      conn
+      |> put_status(:created)
+      |> put_resp_header("location", Routes.canned_response_path(conn, :show, canned_response))
+      |> render("show.json", canned_response: canned_response)
+    end
+  end
+
+  @spec show(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def show(conn, %{"id" => id}) do
+    canned_response = CannedResponses.get_canned_response!(id)
+
+    with :ok <- is_authorized?(conn.assigns.current_user, canned_response) do
+      render(conn, "show.json", canned_response: canned_response)
+    end
+  end
+
+  @spec update(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def update(conn, %{"id" => id, "content" => content, "name" => name}) do
+    canned_response = CannedResponses.get_canned_response!(id)
+
+    with :ok <- is_authorized?(conn.assigns.current_user, canned_response),
+         {:ok, %CannedResponse{} = canned_response} <-
+           CannedResponses.update_canned_response(canned_response, %{
+             name: name,
+             content: content
+           }) do
+      render(conn, "show.json", canned_response: canned_response)
+    end
+  end
+
+  @spec delete(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def delete(conn, %{"id" => id}) do
+    canned_response = CannedResponses.get_canned_response!(id)
+
+    with :ok <- is_authorized?(conn.assigns.current_user, canned_response),
+         {:ok, %CannedResponse{}} <- CannedResponses.delete_canned_response(canned_response) do
+      send_resp(conn, :no_content, "")
+    end
+  end
+
+  defp is_authorized?(user, canned_response) do
+    cond do
+      user.role == "admin" ->
```

Comment:
sorry the docs on user roles aren't great 😅 but we don't want admin users to have this much access -- admins just exist within accounts (not the whole platform) and have a bit more capabilities that regular users, but in this case the permissions should be the same (i.e. both "users" and "admin" can access the same canned response resources)

Reply:
btw some super tiny nitpicks:
- since elixir supports `?` in method names, prefixing with `is_` is usually unnecessary (though i do it often too, coming from JS/TS)
- it's generally expected that methods that end in `?` return a boolean value (rather than `:ok` vs `:error`)
- assuming we wanted to keep the logic in this method the same, it would be a little bit more idiomatic as a `case` statement:

```elixir
def authorize(user, %CannedResponse{account_id: account_id}) do
  case user do
    %{role: "admin"} -> :ok
    %{account_id: ^account_id} -> :ok
    _ -> {:error, :forbidden}
  end
end
```
we could also do:
```elixir
  def authorize(%User{role: "admin"}, _), do: :ok

  def authorize(%User{account_id: account_id}, %CannedResponse{account_id: account_id}),
    do: :ok

  def authorize(_, _), do: {:error, :forbidden}
```
(which is basically just syntatic sugar for the case statement above)

Reply:
thanks for this review, i appreciate it. don't worry about offending with nitpicks 🙂